In [2]:
import pandas as pd
import numpy as np
import warnings
import os

warnings.filterwarnings("ignore")

pd.set_option("display.max_columns", None)
pd.set_option("display.width", 1000)

print("Libraries imported successfully.")

Libraries imported successfully.


In [3]:
print("Current folder:")
print(os.getcwd())

Current folder:
C:\Users\KIIT0001\OneDrive\Desktop\credit-risk-intelligence\notebooks


In [4]:
print("Files in raw folder:")
print(os.listdir("../data/raw"))

Files in raw folder:
['.ipynb_checkpoints', 'accepted_2007_to_2018Q4.csv']


In [6]:
df = pd.read_csv(
    "../data/raw/accepted_2007_to_2018Q4.csv",
    low_memory=False
)

print("Dataset shape:", df.shape)
df.head()

Dataset shape: (2260701, 151)


,id,member_id,loan_amnt,funded_amnt,funded_amnt_inv,term,int_rate,installment,grade,sub_grade,emp_title,emp_length,home_ownership,annual_inc,verification_status,issue_d,loan_status,pymnt_plan,url,desc,purpose,title,zip_code,addr_state,dti,delinq_2yrs,earliest_cr_line,fico_range_low,fico_range_high,inq_last_6mths,mths_since_last_delinq,mths_since_last_record,open_acc,pub_rec,revol_bal,revol_util,total_acc,initial_list_status,out_prncp,out_prncp_inv,total_pymnt,total_pymnt_inv,total_rec_prncp,total_rec_int,total_rec_late_fee,recoveries,collection_recovery_fee,last_pymnt_d,last_pymnt_amnt,next_pymnt_d,last_credit_pull_d,last_fico_range_high,last_fico_range_low,collections_12_mths_ex_med,mths_since_last_major_derog,policy_code,application_type,annual_inc_joint,dti_joint,verification_status_joint,acc_now_delinq,tot_coll_amt,tot_cur_bal,open_acc_6m,open_act_il,open_il_12m,open_il_24m,mths_since_rcnt_il,total_bal_il,il_util,open_rv_12m,open_rv_24m,max_bal_bc,all_util,total_rev_hi_lim,inq_fi,total_cu_tl,inq_last_12m,acc_open_past_24mths,avg_cur_bal,bc_open_to_buy,bc_util,chargeoff_within_12_mths,delinq_amnt,mo_sin_old_il_acct,mo_sin_old_rev_tl_op,mo_sin_rcnt_rev_tl_op,mo_sin_rcnt_tl,mort_acc,mths_since_recent_bc,mths_since_recent_bc_dlq,mths_since_recent_inq,mths_since_recent_revol_delinq,num_accts_ever_120_pd,num_actv_bc_tl,num_actv_rev_tl,num_bc_sats,num_bc_tl,num_il_tl,num_op_rev_tl,num_rev_accts,num_rev_tl_bal_gt_0,num_sats,num_tl_120dpd_2m,num_tl_30dpd,num_tl_90g_dpd_24m,num_tl_op_past_12m,pct_tl_nvr_dlq,percent_bc_gt_75,pub_rec_bankruptcies,tax_liens,tot_hi_cred_lim,total_bal_ex_mort,total_bc_limit,total_il_high_credit_limit,revol_bal_joint,sec_app_fico_range_low,sec_app_fico_range_high,sec_app_earliest_cr_line,sec_app_inq_last_6mths,sec_app_mort_acc,sec_app_open_acc,sec_app_revol_util,sec_app_open_act_il,sec_app_num_rev_accts,sec_app_chargeoff_within_12_mths,sec_app_collections_12_mths_ex_med,sec_app_mths_since_last_major_derog,hardship_flag,hardship_type,hardship_reason,hardship_status,deferral_term,hardship_amount,hardship_start_date,hardship_end_date,payment_plan_start_date,hardship_length,hardship_dpd,hardship_loan_status,orig_projected_additional_accrued_interest,hardship_payoff_balance_amount,hardship_last_payment_amount,disbursement_method,debt_settlement_flag,debt_settlement_flag_date,settlement_status,settlement_date,settlement_amount,settlement_percentage,settlement_term
0,68407277,NaN,3600.0,3600.0,3600.0,36 months,13.99,123.03,C,C4,leadman,10+ years,MORTGAGE,55000.0,Not Verified,Dec-2015,Fully Paid,n,https://lendingclub.com/browse/loanDetail.acti...,NaN,debt_consolidation,Debt consolidation,190xx,PA,5.91,0.0,Aug-2003,675.0,679.0,1.0,30.0,NaN,7.0,0.0,2765.0,29.7,13.0,w,0.00,0.00,4421.723917,4421.72,3600.00,821.72,0.0,0.0,0.0,Jan-2019,122.67,NaN,Mar-2019,564.0,560.0,0.0,30.0,1.0,Individual,NaN,NaN,NaN,0.0,722.0,144904.0,2.0,2.0,0.0,1.0,21.0,4981.0,36.0,3.0,3.0,722.0,34.0,9300.0,3.0,1.0,4.0,4.0,20701.0,1506.0,37.2,0.0,0.0,148.0,128.0,3.0,3.0,1.0,4.0,69.0,4.0,69.0,2.0,2.0,4.0,2.0,5.0,3.0,4.0,9.0,4.0,7.0,0.0,0.0,0.0,3.0,76.9,0.0,0.0,0.0,178050.0,7746.0,2400.0,13734.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,N,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN
1,68355089,NaN,24700.0,24700.0,24700.0,36 months,11.99,820.28,C,C1,Engineer,10+ years,MORTGAGE,65000.0,Not Verified,Dec-2015,Fully Paid,n,https://lendingclub.com/browse/loanDetail.acti...,NaN,small_business,Business,577xx,SD,16.06,1.0,Dec-1999,715.0,719.0,4.0,6.0,NaN,22.0,0.0,21470.0,19.2,38.0,w,0.00,0.00,25679.660000,25679.66,24700.00,979.66,0.0,0.0,0.0,Jun-2016,926.35,NaN,Mar-2019,699.0,695.0,0.0,NaN,1.0,Individual,NaN,NaN,NaN,0.0,0.0,204396.0,1.0,1.0,0.0,1.0,19.0,18005.0,73.0,2.0,3.0,6472.0,29.0,111800.0,0.0,0.0,6.0,4.0,9733.0,57830.0,27.1,0.0,0.0,113.0,192.0,2.0,2.0,4.0,2.0,NaN,0.0,6.0,0.0,5.0,5.0,13.0,17.0,6.0,20.0,27.0,5.0,22.0,0.0,0.0,0.0,2.0,97.4,7.7,0.0,0.0,314017.0,39475.0,79300.0,24667.0,NaN,NaN,NaN,NaN,NaN,N

In [7]:
df["loan_status"].value_counts()

loan_status
Fully Paid                                             1076751
Current                                                 878317
Charged Off                                             268559
Late (31-120 days)                                       21467
In Grace Period                                           8436
Late (16-30 days)                                         4349
Does not meet the credit policy. Status:Fully Paid        1988
Does not meet the credit policy. Status:Charged Off        761
Default                                                     40
Name: count, dtype: int64

In [8]:
bad_statuses = [
    "Charged Off",
    "Default",
    "Does not meet the credit policy. Status:Charged Off"
]

good_statuses = [
    "Fully Paid",
    "Does not meet the credit policy. Status:Fully Paid"
]

df_model = df[df["loan_status"].isin(bad_statuses + good_statuses)].copy()

df_model["target"] = df_model["loan_status"].isin(bad_statuses).astype(int)

print("Final modeling dataset shape:", df_model.shape)

print("\nTarget count:")
print(df_model["target"].value_counts())

print("\nTarget percentage:")
print(df_model["target"].value_counts(normalize=True).mul(100).round(2))

Final modeling dataset shape: (1348099, 152)

Target count:
target
0    1078739
1     269360
Name: count, dtype: int64

Target percentage:
target
0    80.02
1    19.98
Name: proportion, dtype: float64


### Target Variable Definition

The target variable was created from the `loan_status` column. Loans with `Fully Paid` and `Does not meet the credit policy. Status:Fully Paid` were labelled as **0**, representing non-default borrowers.

Loans with `Charged Off`, `Default`, and `Does not meet the credit policy. Status:Charged Off` were labelled as **1**, representing default borrowers.

Active or uncertain loans such as `Current`, `Late`, and `In Grace Period` were excluded because their final repayment outcome was not confirmed.

After filtering, the final modeling dataset contains **1,348,099 records** and **152 columns**. The target distribution is **80.02% non-default borrowers** and **19.98% default borrowers**.

In [9]:
dict_df = pd.DataFrame({
    "Column": df.columns,
    "Data_Type": df.dtypes.values.astype(str),
    "Non_Null_Count": df.notnull().sum().values,
    "Null_Percent": (df.isnull().mean() * 100).round(2).values,
    "Unique_Values": df.nunique().values,
    "Sample_Value": [
        str(df[c].dropna().iloc[0]) if df[c].notnull().any() else "ALL NULL"
        for c in df.columns
    ],
    "Description": "",
    "ML_Importance": ""
})

dict_df.to_excel("../report/data_dictionary.xlsx", index=False)

print("Data dictionary created successfully.")

Data dictionary created successfully.


### Data Dictionary Note

The data dictionary was created from the original raw LendingClub dataset (`df`) to document all available columns, data types, missing value percentages, unique values, and sample values. The filtered modeling dataset (`df_model`) was created separately only for defining the target variable and future modeling work.

In [11]:
print("Main folder files:")
print(os.listdir(".."))

print("\nReport folder files:")
print(os.listdir("../report"))

Main folder files:
['dashboard', 'data', 'models', 'notebooks', 'README.md', 'report', 'scripts']

Report folder files:
['.ipynb_checkpoints', 'data_dictionary.xlsx', 'problem_statement.md']


In [12]:
print(os.listdir(".."))

['.gitignore', '.ipynb_checkpoints', 'dashboard', 'data', 'models', 'notebooks', 'problem_statement.md', 'README.md', 'report', 'scripts', 'untitled.txt']


In [13]:
print(os.listdir(".."))

['.gitignore', '.ipynb_checkpoints', 'dashboard', 'data', 'models', 'notebooks', 'problem_statement.md', 'README.md', 'report', 'requirements.txt', 'scripts']
